pkg/reconciler/taskrun/resources/apply.go

```diff
@@ -46,14 +46,73 @@ var (
 	}
 )
 
-// ApplyParameters applies the params from a TaskRun.Input.Parameters to a TaskSpec
-func ApplyParameters(ctx context.Context, spec *v1.TaskSpec, tr *v1.TaskRun, defaults ...v1.ParamSpec) *v1.TaskSpec {
+// applyStepActionParameters applies the params from a TaskRun.Input.Parameters to a TaskSpec
+func applyStepActionParameters(ctx context.Context, step *v1.Step, spec *v1.TaskSpec, tr *v1.TaskRun, stepParams v1.Params, defaults []v1.ParamSpec) *v1.Step {
+	if stepParams != nil {
+		stringR, arrayR, objectR := getTaskParameters(ctx, spec, tr, spec.Params...)
+		stepParams = stepParams.ReplaceVariables(stringR, arrayR, objectR)
+	}
+	stringReplacements := map[string]string{}
+	arrayReplacements := map[string][]string{}
+	objectReplacements := map[string]map[string]string{}
+
+	// Set all the default stringReplacements
+	for _, p := range defaults {
+		if p.Default != nil {
+			switch p.Default.Type {
+			case v1.ParamTypeArray:
+				for _, pattern := range paramPatterns {
+					for i := 0; i < len(p.Default.ArrayVal); i++ {
+						stringReplacements[fmt.Sprintf(pattern+"[%d]", p.Name, i)] = p.Default.ArrayVal[i]
+					}
+					arrayReplacements[fmt.Sprintf(pattern, p.Name)] = p.Default.ArrayVal
+				}
+			case v1.ParamTypeObject:
+				for _, pattern := range paramPatterns {
+					objectReplacements[fmt.Sprintf(pattern, p.Name)] = p.Default.ObjectVal
+				}
+				for k, v := range p.Default.ObjectVal {
+					stringReplacements[fmt.Sprintf(objectIndividualVariablePattern, p.Name, k)] = v
+				}
+			case v1.ParamTypeString:
+				fallthrough
+			default:
+				for _, pattern := range paramPatterns {
+					stringReplacements[fmt.Sprintf(pattern, p.Name)] = p.Default.StringVal
+				}
+			}
+		}
+	}
+
+	// Set and overwrite params with the ones from the Step
+	stepStrings, stepArrays, stepObjects := paramsFromStep(ctx, stepParams)
```

Comment:
I think `paramsFromStep` and current `paramsFromTaskRun` can be merged into one if we don't need the
`stepObjects` here (looks like it is not used because objectReplacements is not used) and change  `paramsFromTaskRun` to accept the v1.Params instead of a `taskrun`

Reply:
paramsFromTaskRun needs to output objectParams but paramsFromStep does not.

Reply:
cleaned up all around. PTAL.